Switch over to a `Makefile`

joneshf-cn opened the pull request:
##### What it does

As discussed on slack, we're going to try using a `Makefile` for building PS stuff. This is mostly a direct port of the npm scripts. It does set up dependencies automatically though. I didn't want to spend too much time on this, so it's still pretty naïve. It could be made a bit better by using `purs` directly instead of going through `pulp` still.

More than happy to explain more motivation if it's not clear from the commit/slack conversation.

##### Review asks

Are we cool with moving to a `Makefile`?

Requesting a review from Dave (since you're working a bunch in this repo) and Forest (since we're in this escapade together).

joneshf-cn committed:
Switch over to a `Makefile`

Make tends to handle building of a project a bit better than
npm scripts as it grows in complexity.

We were just starting to get to that point.
We wanted to use `psa` to turn warnings into errors,
but `pulp` will mess up if we try to pass flags to `psa` all the time.
So we had to split it up into two scripts.
And we could only use the one script after the other had run.

There is a bit of work hiding in npm scripts though,
we have to always have things installed an up to date.
When first starting on the project, we have to remember to:
1. `npm install`
2. `npx bower install`
3. `npm run build`

When coming back to the project,
it's very possible the dependencies will change
and will require the same set of commands.

Or maybe the commands to get up to date have changed,
and we're using `psc-package` or `spago` or something.

Switching to a `Makefile` should be less painful in the long run.
Hopefully, people can come in, type `make` and be good to go.
No need to remember what order of commands to run,
no need to remember what commands to run, just `make`.

joneshf-cn committed:
Remove the npm scripts altogether

Might as well mitigate the npm scripts getting out of sync with make.
Nothing worse than having two ways of doing the same thing,
and one of them doesn't work.

Go big or go home, they say.

davezuch commented on Makefile:
```diff
@@ -0,0 +1,76 @@
+
+.PHONY: watch
+watch: $(BOWER_COMPONENTS) $(NODE_MODULES)
+	npx pulp --watch build --to dist/main.js
```
I've always had problems with this watch, sometimes it rebuilds twice (on any of our EC2's at least, didn't experience locally), sometimes doesn't build at all. I saw some people in Slack (you may have been one of them) suggest `watch-exec` instead, and that's worked beautifully. It is an extra dependency, but might be worth using instead

foresttoney replied:
I have also seen this behavior - I didn't know there was something I could install to make my life better.

joneshf-cn replied:
It looks simple enough to add, so I'm down to give it a shot!

davezuch reviewed (approved):
Sorry it took me a while to review, I wasn't very familiar with `make`, so had to do a little reading when I didn't get something. This looks dope though 👍

foresttoney reviewed (approved):
This all looks good to me

joneshf-cn commented:
> I wasn't very familiar with `make`, so had to do a little reading when I didn't get something.

I'm more than happy to add whatever caught you up to the `Makefile` as documentation. It's likely someone else will get bit by it as well. What stuff should we add?

joneshf-cn committed:
Use `watch-exec` instead of `pulp --watch`

We run into an issue with `pulp --watch`:
on some machines, it will double build when a file changes.
Some of us have been using `watch-exec` instead,
and it's not had the same issue as `pulp --watch`.

We change the `Makefile` to use `watch-exec`,
so everyone is on the same page.

davezuch commented:
> I'm more than happy to add whatever caught you up to the Makefile as documentation. It's likely someone else will get bit by it as well. What stuff should we add?

The comments and links in your Makefile were enough for me. Just taking the time to read the more detailed explanations and parse the examples is what did it. If anything I might explain or link to documentation for a few more things like the `.DEFAULT_GOAL` and `.PHONY` keywords, but that's all available when you follow any of the links

joneshf-cn committed:
Explain a bit more about make

There are a few things we're using that aren't well known/explained
(like `.PHONY` and friends).
We can add more documentation to help people better understand this file.

joneshf-cn commented:
Expanded on the docs. Lemme know if there's more we should do.
